File: arckit-plugin/hooks/validate-wardley-math.py
#!/usr/bin/env python3
"""
ArcKit Wardley Map Math Validation

Validates a generated Wardley Map document for consistency:
  1. Stage-evolution alignment (Component Inventory tables)
  2. Coordinate range validation (all values in [0.00, 1.00])
  3. OWM syntax consistency (wardley code block vs Component Inventory)

Input (stdin):  JSON { stop_hook_active, ... }
Output (stdout): JSON with "decision": "block" and "reason" on failure, or empty on success
Exit codes:      0 always (block via JSON decision, not exit code)
"""

import json
import os
import re
import sys
import time
from pathlib import Path


def evolution_to_stage(evo):
    """Classify evolution value into expected stage."""
    evo = float(evo)
    if evo < 0.25:
        return "Genesis"
    elif evo < 0.50:
        return "Custom"
    elif evo < 0.75:
        return "Product"
    else:
        return "Commodity"


def main():
    try:
        raw = sys.stdin.read()
        if not raw or not raw.strip():
            sys.exit(0)
        data = json.loads(raw)
    except (json.JSONDecodeError, ValueError, EOFError):
        sys.exit(0)

    # If this is already a re-fire after a block, allow stop to prevent infinite loops
    if data.get("stop_hook_active", False):
        sys.exit(0)

    # --- Find most recently modified WARD file ---
    ward_file = None
    newest_mtime = 0
    now = time.time()
    max_age = 300  # 5 minutes

    # Walk projects/*/wardley-maps/ looking for ARC-*-WARD-*.md
    projects_dir = Path("projects")
    if projects_dir.is_dir():
        for project_dir in sorted(projects_dir.iterdir()):
            wm_dir = project_dir / "wardley-maps"
            if not wm_dir.is_dir():
                continue
            for f in wm_dir.iterdir():
                if not f.is_file():
                    continue
                if not f.name.startswith("ARC-") or "-WARD-" not in f.name or not f.name.endswith(".md"):
                    continue
                mtime = f.stat().st_mtime
                age = now - mtime
                if age <= max_age and mtime > newest_mtime:
                    newest_mtime = mtime
                    ward_file = f

    # No recent wardley file found - not a wardley run, allow stop
    if ward_file is None:
        sys.exit(0)

    filename = ward_file.name
    content = ward_file.read_text(encoding="utf-8", errors="replace")
    lines = content.splitlines()

    errors = []
    stage_errors = []
    owm_errors = []

    # Regex for table rows: | Component | 0.XX | 0.XX | Stage | ... |
    table_row_re = re.compile(
        r"^\|\s*([^|]+?)\s*\|\s*(\d+\.\d+)\s*\|\s*(\d+\.\d+)\s*\|\s*(Genesis|Custom|Product|Commodity)\s*\|"
    )

    # --- Check 1 & 2: Stage-evolution alignment and coordinate range ---
    table_vis = {}
    table_evo = {}

    for line in lines:
        m = table_row_re.match(line)
        if not m:
            continue
        comp = m.group(1).strip()
        vis = m.group(2)
        evo = m.group(3)
        stage = m.group(4)

        # Skip template placeholder rows
        if "{" in comp or comp == "Component":
            continue

        # Check stage-evolution alignment
        expected = evolution_to_stage(evo)
        if stage != expected:
            stage_errors.append(
                f"- '{comp}' has evolution {evo} but Stage is '{stage}' (expected '{expected}')"
            )

        # Check coordinate range
        vis_f = float(vis)
        evo_f = float(evo)
        if vis_f < 0.0 or vis_f > 1.0:
            errors.append(f"- '{comp}' has visibility {vis} outside valid range [0.00, 1.00]")
        if evo_f < 0.0 or evo_f > 1.0:
            errors.append(f"- '{comp}' has evolution {evo} outside valid range [0.00, 1.00]")

        # Store for cross-reference
        table_vis[comp] = vis
        table_evo[comp] = evo

    # --- Check 3: OWM syntax consistency ---
    owm_vis = {}
    owm_evo = {}
    in_wardley = False
    component_re = re.compile(r"^\s*component\s+(.+?)\s+\[\s*([0-9.]+)\s*,\s*([0-9.]+)\s*\]")

    for line in lines:
        if re.match(r"^\s*```wardley", line):
            in_wardley = True
            continue
        if in_wardley and re.match(r"^\s*```", line):
            in_wardley = False
            continue
        if in_wardley:
            cm = component_re.match(line)
            if cm:
                comp_name = cm.group(1).strip()
                owm_v = cm.group(2)
                owm_e = cm.group(3)
                owm_vis[comp_name] = owm_v
                owm_evo[comp_name] = owm_e

    # Cross-reference OWM coordinates vs table coordinates
    for comp_name in owm_vis:
        if comp_name in table_vis:
            t_vis = table_vis[comp_name]
            t_evo = table_evo[comp_name]
            o_vis = owm_vis[comp_name]
            o_evo = owm_evo[comp_name]

            if o_vis != t_vis or o_evo != t_evo:
                owm_errors.append(
                    f"- '{comp_name}' is [{o_vis}, {o_evo}] in OWM but [{t_vis}, {t_evo}] in Component Inventory"
                )

    # --- Build error report ---
    report_parts = []

    if stage_errors:
        report_parts.append("**Stage-Evolution Mismatches:**\n" + "\n".join(stage_errors))

    if errors:
        report_parts.append("**Coordinate Range Errors:**\n" + "\n".join(errors))

    if owm_errors:
        report_parts.append("**OWM Coordinate Mismatches:**\n" + "\n".join(owm_errors))

    if report_parts:
        report = "\n\n".join(report_parts)
        reason = f"Wardley Map validation errors in {filename}:\n\n{report}\n\nFix these errors in the document, then stop again."
        print(json.dumps({"decision": "block", "reason": reason}))

    sys.exit(0)


if __name__ == "__main__":
    main()

File: arckit-claude/hooks/validate-wardley-math.mjs
#!/usr/bin/env node
/**
 * ArcKit PreToolUse (Write) Hook — Wardley Map Math Validation
 *
 * Validates a Wardley Map document about to be written for consistency:
 *   1. Stage-evolution alignment (Component Inventory tables)
 *   2. Coordinate range validation (all values in [0.00, 1.00])
 *   3. OWM syntax consistency (wardley/owm code block vs Component Inventory)
 *   4. Mermaid wardley-beta syntax (unquoted bare-digit tokens break rendering)
 *
 * Hook Type: PreToolUse
 * Matcher: Write
 * Scoped via an `if:` rule in hooks.json that matches Write calls under
 * projects/<id>/wardley-maps/ so it only runs for ARC Wardley Map artefacts.
 *
 * Input (stdin):  JSON { tool_name, tool_input: { file_path, content }, ... }
 * Output (stdout): JSON { decision: "block", reason } on failure; empty on pass.
 * Exit code:       0 in all cases (block via JSON decision so the reason is fed
 *                  back to the model instead of producing a hard permission error).
 */

import { readFileSync } from 'node:fs';
import { basename } from 'node:path';

function evolutionToStage(evo) {
  const val = parseFloat(evo);
  if (val < 0.25) return 'Genesis';
  if (val < 0.50) return 'Custom';
  if (val < 0.75) return 'Product';
  return 'Commodity';
}

// --- Main ---
let raw = '';
try {
  raw = readFileSync(0, 'utf8');
} catch {
  process.exit(0);
}
if (!raw || !raw.trim()) process.exit(0);

let data;
try {
  data = JSON.parse(raw);
} catch {
  process.exit(0);
}

const filePath = (data.tool_input || {}).file_path || '';
const content = (data.tool_input || {}).content || '';

// Defense-in-depth: the `if:` rule in hooks.json already narrows this,
// but skip cleanly if either field is empty or the path isn't a Wardley artefact.
if (!filePath || !content) process.exit(0);
if (!filePath.includes('/wardley-maps/')) process.exit(0);

const filename = basename(filePath);
const contentLines = content.split('\n');

const errors = [];
const stageErrors = [];
const owmErrors = [];

// Regex for table rows: | Component | 0.XX | 0.XX | Stage | ... |
const tableRowRe = /^\|\s*([^|]+?)\s*\|\s*(\d+\.\d+)\s*\|\s*(\d+\.\d+)\s*\|\s*(Genesis|Custom|Product|Commodity)\s*\|/;

// --- Check 1 & 2: Stage-evolution alignment and coordinate range ---
const tableVis = {};
const tableEvo = {};

for (const line of contentLines) {
  const m = line.match(tableRowRe);
  if (!m) continue;
  const comp = m[1].trim();
  const vis = m[2];
  const evo = m[3];
  const stage = m[4];

  // Skip template placeholder rows
  if (comp.includes('{') || comp === 'Component') continue;

  // Check stage-evolution alignment
  const expected = evolutionToStage(evo);
  if (stage !== expected) {
    stageErrors.push(`- '${comp}' has evolution ${evo} but Stage is '${stage}' (expected '${expected}')`);
  }

  // Check coordinate range
  const visF = parseFloat(vis);
  const evoF = parseFloat(evo);
  if (visF < 0.0 || visF > 1.0) {
    errors.push(`- '${comp}' has visibility ${vis} outside valid range [0.00, 1.00]`);
  }
  if (evoF < 0.0 || evoF > 1.0) {
    errors.push(`- '${comp}' has evolution ${evo} outside valid range [0.00, 1.00]`);
  }

  // Store for cross-reference
  tableVis[comp] = vis;
  tableEvo[comp] = evo;
}

// --- Check 3: OWM syntax consistency ---
// Accepts both ```wardley and ```owm fence aliases (OnlineWardleyMaps uses both).
const owmVis = {};
const owmEvo = {};
let inWardley = false;
const componentRe = /^\s*component\s+(.+?)\s+\[\s*([0-9.]+)\s*,\s*([0-9.]+)\s*\]/;
const owmFenceOpenRe = /^\s*```(?:wardley|owm)\b/;

for (const line of contentLines) {
  if (owmFenceOpenRe.test(line)) {
    inWardley = true;
    continue;
  }
  if (inWardley && /^\s*```/.test(line)) {
    inWardley = false;
    continue;
  }
  if (inWardley) {
    const cm = line.match(componentRe);
    if (cm) {
      const compName = cm[1].trim();
      owmVis[compName] = cm[2];
      owmEvo[compName] = cm[3];

      // Also enforce OWM coordinate range — the plane is unit square [0, 1].
      const oVisF = parseFloat(cm[2]);
      const oEvoF = parseFloat(cm[3]);
      if (oVisF < 0.0 || oVisF > 1.0) {
        errors.push(`- '${compName}' OWM block has visibility ${cm[2]} outside valid range [0.00, 1.00]`);
      }
      if (oEvoF < 0.0 || oEvoF > 1.0) {
        errors.push(`- '${compName}' OWM block has evolution ${cm[3]} outside valid range [0.00, 1.00]`);
      }
    }
  }
}

// Cross-reference OWM coordinates vs table coordinates
for (const compName of Object.keys(owmVis)) {
  if (compName in tableVis) {
    const tVis = tableVis[compName];
    const tEvo = tableEvo[compName];
    const oVis = owmVis[compName];
    const oEvo = owmEvo[compName];

    if (oVis !== tVis || oEvo !== tEvo) {
      owmErrors.push(
        `- '${compName}' is [${oVis}, ${oEvo}] in OWM but [${tVis}, ${tEvo}] in Component Inventory`
      );
    }
  }
}

// --- Check 4: Mermaid wardley-beta syntax (unquoted bare-digit tokens) ---
//
// The wardley-beta parser tokenises bare numeric words (like `2031`, `27001`)
// as numeric literals, which breaks rendering with errors such as:
//   "Parse error on line N, column M: Expecting token of type '[' but found '2031'."
// Any name containing a whitespace-separated pure-digit word MUST be quoted
// everywhere it appears — component/anchor declarations, both sides of `->`
// edges, `evolve` targets, and `pipeline` parents.

const mermaidErrors = [];
let inMermaidBlock = false;
let inMermaidWardley = false;

function extractNameZones(line) {
  // component NAME [v, e] (decorators)
  let m = line.match(/^\s*component\s+(.+?)\s*\[/);
  if (m) return [m[1]];
  // anchor NAME [v, e]
  m = line.match(/^\s*anchor\s+(.+?)\s*\[/);
  if (m) return [m[1]];
  // evolve NAME 0.45 (trailing number is the target evolution coord)
  m = line.match(/^\s*evolve\s+(.+?)\s+[0-9.]+\s*$/);
  if (m) return [m[1]];
  // pipeline NAME [v1, v2]? (brackets optional)
  m = line.match(/^\s*pipeline\s+(.+?)(?:\s*\[|\s*$)/);
  if (m) return [m[1]];
  // edge: NAME -> NAME (single arrow, two zones)
  if (line.includes('->') && !/^\s*(?:component|anchor|evolve|pipeline|note|annotation|annotations|title|size|style)\b/.test(line)) {
    const parts = line.split('->');
    if (parts.length === 2) return [parts[0], parts[1]];
  }
  return [];
}

function bareDigitWords(nameZone) {
  // Strip quoted substrings and decorator groups so we only inspect unquoted words
  const stripped = nameZone
    .replace(/"[^"]*"/g, '')
    .replace(/\([^)]*\)/g, '')
    .trim();
  if (!stripped) return [];
  return stripped.split(/\s+/).filter((w) => /^\d+$/.test(w));
}

for (let i = 0; i < contentLines.length; i++) {
  const line = contentLines[i];
  const trimmed = line.trim();

  if (/^```mermaid\b/.test(trimmed)) {
    inMermaidBlock = true;
    inMermaidWardley = false;
    continue;
  }
  if (inMermaidBlock && /^```/.test(trimmed)) {
    inMermaidBlock = false;
    inMermaidWardley = false;
    continue;
  }
  if (!inMermaidBlock) continue;

  if (!inMermaidWardley) {
    if (/^wardley-beta\b/.test(trimmed)) inMermaidWardley = true;
    continue;
  }

  for (const zone of extractNameZones(line)) {
    const bad = bareDigitWords(zone);
    if (bad.length > 0) {
      mermaidErrors.push(
        `- Line ${i + 1}: unquoted name '${zone.trim()}' contains bare numeric word(s) '${bad.join("', '")}' — wrap the whole name in double quotes everywhere it appears (declaration, both sides of '->', 'evolve' targets, 'pipeline' parents)`
      );
    }
  }
}

// --- Build error report ---
const reportParts = [];

if (stageErrors.length > 0) {
  reportParts.push('**Stage-Evolution Mismatches:**\n' + stageErrors.join('\n'));
}
if (errors.length > 0) {
  reportParts.push('**Coordinate Range Errors:**\n' + errors.join('\n'));
}
if (owmErrors.length > 0) {
  reportParts.push('**OWM Coordinate Mismatches:**\n' + owmErrors.join('\n'));
}
if (mermaidErrors.length > 0) {
  reportParts.push('**Mermaid wardley-beta Syntax Errors (bare numeric tokens break rendering):**\n' + mermaidErrors.join('\n'));
}

if (reportParts.length > 0) {
  const report = reportParts.join('\n\n');
  const reason = `Wardley Map validation errors in ${filename}:\n\n${report}\n\nFix these errors and re-issue the Write.`;
  console.log(JSON.stringify({ decision: 'block', reason }));
}

process.exit(0);
